vault/request_handling.go

```diff
@@ -868,24 +867,34 @@ func (c *Core) handleRequest(ctx context.Context, req *logical.Request) (retResp
 			return nil, nil, ErrInternalError
 		}
 
-		resp.Auth.TokenPolicies = policyutil.SanitizePolicies(resp.Auth.Policies, policyutil.DoNotAddDefaultPolicy)
-		switch resp.Auth.TokenType {
-		case logical.TokenTypeBatch:
-		case logical.TokenTypeService:
-			if err := c.expiration.RegisterAuth(ctx, &logical.TokenEntry{
-				TTL:         auth.TTL,
-				Policies:    auth.TokenPolicies,
-				Path:        resp.Auth.CreationPath,
-				NamespaceID: ns.ID,
-			}, resp.Auth); err != nil {
-				// Best-effort clean up on error, so we log the cleanup error as
-				// a warning but still return as internal error.
-				if err := c.tokenStore.revokeOrphan(ctx, resp.Auth.ClientToken); err != nil {
-					c.logger.Warn("failed to clean up token lease during auth/token/ request", "request_path", req.Path, "error", err)
+		// We skip expiration manager registration for token renewal since it
+		// does not need to be re-registered
+		if strings.HasPrefix(req.Path, "auth/token/renew") {
+			// We build the "policies" list to be returned by starting with
+			// token policies, and add identity policies right after this
+			// conditional
+			resp.Auth.Policies = policyutil.SanitizePolicies(resp.Auth.TokenPolicies, policyutil.DoNotAddDefaultPolicy)
```

Comment:
Why do we assign to Policies here instead of TokenPolicies as in the other branch?

Assuming my intuition is correct - that the assignment should be the same regardless of whether or not this is a renewal -  I think the code might be a bit clearer if you do the assignment as a first step, then extend the following switch statement, like:

```
resp.Auth.TokenPolicies = policyutil.SanitizePolicies(resp.Auth.Policies, policyutil.DoNotAddDefaultPolicy)
switch {
case strings.HasPrefix(req.Path, "auth/token/renew"): // do nothing
case resp.Auth.TokenType == logical.TokenTypeBatch: // do nothing
case resp.Auth.TokenType == logical.TokenTypeService: // do registration
}
```

Reply:
This doesn't work because `resp.Auth.Policies` returned by a renewal response after routing is empty (the token policies are returned under `resp.Auth.TokenPolicies`), and hence why we need to add them to `resp.Auth.Policies` and not the other way around.

Reply:
I wish I understood why it's like that, but my timeboxed foray into the code for answers didn't bear fruit.  Anyway, I don't think this will break anything, and it solves the existing problem, so I'll let it go.

Reply:
I dug a bit further to try to understand it a bit more as well. The `Auth` object in `resp.Auth` comes from the associated lease entry in the token, which is assigned in `renewAuthEntry` that we hit during the renewal request flow. It looks like the `policies` field omits the default policy, and thus either 1) does not contain the full set of token policies or 2) stores a `null` object for that field if default is the only policy (as seen by the examples below when reading the lease entry in storage). That inconsistency may or may not need to be addressed separately, but in any case using `token_policies` as the starting set for token policies is the more reliable way to build `resp.Auth.Policies`.

User bob on userpass with no policy attached, just the default:
```
  "auth": {
    "lease": 2764798000000000,
    "max_ttl": 0,
    "renewable": true,
    "internal_data": null,
    "display_name": "userpass-bob",
    "policies": null,
    "token_policies": [
      "default"
    ],
...
```

User bob with "test" policy attached:
```
  "auth": {
    "lease": 2764798000000000,
    "max_ttl": 0,
    "renewable": true,
    "internal_data": null,
    "display_name": "userpass-bob",
    "policies": [
      "test"
    ],
    "token_policies": [
      "default",
      "test"
    ],
```